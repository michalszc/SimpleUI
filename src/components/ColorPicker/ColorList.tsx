import React, { FC } from "react";
import * as CSS from "csstype";
import styled from "styled-components";
import ColorSelector from "./ColorSelector";
import Styles from "../../utils/styles";
import { IStyledTheme } from "../../utils/theme/theme";

export interface ColorListProps {
    colors: Array<CSS.DataType.Color>;
    shape: 'circle' | 'square' | 'roundedSquare';
    size: [number, number];
    setColor: (value: CSS.DataType.Color) => void;
}

interface IList extends IStyledTheme {
    size: [number, number];
}

const List = styled.div<IList>`
    padding: ${ ({ theme }) => theme.sizes[3] };
    display: grid;
    gap: ${ ({ theme }) => theme.sizes[1.5] };
    grid-template-rows: ${ ({ size }) => 'auto '.repeat(size.at(0)!)};
    grid-template-columns: ${ ({ size }) => 'auto '.repeat(size.at(1)!)};
`;

const ColorList: FC<ColorListProps> = ({
    colors, shape, size, setColor
}) => (
    <Styles>
        <List size={size}>
            {
                colors
                    .map((color: CSS.DataType.Color, i: number) =>
                        <ColorSelector key={i} color={color} shape={shape} setColor={setColor} />
                    )
            }
        </List>
    </Styles>
);

export default ColorList;
