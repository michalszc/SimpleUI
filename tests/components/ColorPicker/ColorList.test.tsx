import React from "react";
import { fireEvent, render } from "@testing-library/react";
import ColorList from "../../../src/components/ColorPicker/ColorList";
import { DataType } from "csstype";

function setupEyeDropper() {
    Object.defineProperty(window, 'EyeDropper', {
        writable: true,
        value: jest.fn().mockImplementation(() => ({})),
    });
}

describe("ColorList", () => {
    const mockFn = jest.fn();
    const colors: DataType.Color[] = [
        "#b2dfdb",
        "#ffcdd2",
        "#b3e5fc",
        "#d7ccc8"
    ];
    test("should properly render ColorList component - shape circle", () => {
        const container = render(<ColorList colors={colors} shape={'circle'} size={[2,2]} eyeDropper={false} setColor={mockFn} />);
        expect(container).toMatchSnapshot();
    });
    test("should properly render ColorList component - shape square", () => {
        const container = render(<ColorList colors={colors} shape={'square'} size={[2,2]} eyeDropper={false} setColor={mockFn} />);
        expect(container).toMatchSnapshot();
    });
    test("should properly render ColorList component - shape roundedSquare", () => {
        const container = render(<ColorList colors={colors} shape={'roundedSquare'} size={[2,2]} eyeDropper={false} setColor={mockFn} />);
        expect(container).toMatchSnapshot();
    });
    test("should properly render ColorList component with EyeDropper", () => {
        setupEyeDropper();
        const { container, getByTestId } = render(<ColorList colors={colors} shape={'circle'} size={[2,2]} eyeDropper={true} setColor={mockFn} />);
        const eyeDropper = getByTestId('test-eyedropper');
        expect(container).toMatchSnapshot();
        expect(eyeDropper).toBeTruthy();
    });
    test("should properly trigger setColor function on click", () => {
        const { getAllByTestId } = render(<ColorList colors={colors} shape={'circle'} size={[2,2]} eyeDropper={true} setColor={mockFn} />);
        const colorSelectors = getAllByTestId('test-colorselector');
        fireEvent.click(colorSelectors[0]);
        expect(mockFn).toBeCalled();
    });
});
