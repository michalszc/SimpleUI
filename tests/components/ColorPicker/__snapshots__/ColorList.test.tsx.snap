// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorList should properly render ColorList component - shape circle 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-dkrFOg bgZgCB simpleui-colorpicker-list"
        id="4fzzzxjylrx"
        mode="light"
      >
        <div
          class="sc-bcXHqe evoygh simpleui-colorpicker-colorselector"
          color="#b2dfdb"
          data-testid="test-colorselector"
          id="4fzzzxjylrx"
          mode="light"
          shape="circle"
          title="#b2dfdb"
        />
        <div
          class="sc-bcXHqe jxHNVE simpleui-colorpicker-colorselector"
          color="#ffcdd2"
          data-testid="test-colorselector"
          id="4fzzzxjylrx"
          mode="light"
          shape="circle"
          title="#ffcdd2"
        />
        <div
          class="sc-bcXHqe ldIywx simpleui-colorpicker-colorselector"
          color="#b3e5fc"
          data-testid="test-colorselector"
          id="4fzzzxjylrx"
          mode="light"
          shape="circle"
          title="#b3e5fc"
        />
        <div
          class="sc-bcXHqe jRrVxh simpleui-colorpicker-colorselector"
          color="#d7ccc8"
          data-testid="test-colorselector"
          id="4fzzzxjylrx"
          mode="light"
          shape="circle"
          title="#d7ccc8"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-dkrFOg bgZgCB simpleui-colorpicker-list"
      id="4fzzzxjylrx"
      mode="light"
    >
      <div
        class="sc-bcXHqe evoygh simpleui-colorpicker-colorselector"
        color="#b2dfdb"
        data-testid="test-colorselector"
        id="4fzzzxjylrx"
        mode="light"
        shape="circle"
        title="#b2dfdb"
      />
      <div
        class="sc-bcXHqe jxHNVE simpleui-colorpicker-colorselector"
        color="#ffcdd2"
        data-testid="test-colorselector"
        id="4fzzzxjylrx"
        mode="light"
        shape="circle"
        title="#ffcdd2"
      />
      <div
        class="sc-bcXHqe ldIywx simpleui-colorpicker-colorselector"
        color="#b3e5fc"
        data-testid="test-colorselector"
        id="4fzzzxjylrx"
        mode="light"
        shape="circle"
        title="#b3e5fc"
      />
      <div
        class="sc-bcXHqe jRrVxh simpleui-colorpicker-colorselector"
        color="#d7ccc8"
        data-testid="test-colorselector"
        id="4fzzzxjylrx"
        mode="light"
        shape="circle"
        title="#d7ccc8"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ColorList should properly render ColorList component - shape roundedSquare 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-dkrFOg bgZgCB simpleui-colorpicker-list"
        id="4fzzzxjylrx"
        mode="light"
      >
        <div
          class="sc-bcXHqe kaVimm simpleui-colorpicker-colorselector"
          color="#b2dfdb"
          data-testid="test-colorselector"
          id="4fzzzxjylrx"
          mode="light"
          shape="roundedSquare"
          title="#b2dfdb"
        />
        <div
          class="sc-bcXHqe junUgb simpleui-colorpicker-colorselector"
          color="#ffcdd2"
          data-testid="test-colorselector"
          id="4fzzzxjylrx"
          mode="light"
          shape="roundedSquare"
          title="#ffcdd2"
        />
        <div
          class="sc-bcXHqe inHKGm simpleui-colorpicker-colorselector"
          color="#b3e5fc"
          data-testid="test-colorselector"
          id="4fzzzxjylrx"
          mode="light"
          shape="roundedSquare"
          title="#b3e5fc"
        />
        <div
          class="sc-bcXHqe bRjGSW simpleui-colorpicker-colorselector"
          color="#d7ccc8"
          data-testid="test-colorselector"
          id="4fzzzxjylrx"
          mode="light"
          shape="roundedSquare"
          title="#d7ccc8"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-dkrFOg bgZgCB simpleui-colorpicker-list"
      id="4fzzzxjylrx"
      mode="light"
    >
      <div
        class="sc-bcXHqe kaVimm simpleui-colorpicker-colorselector"
        color="#b2dfdb"
        data-testid="test-colorselector"
        id="4fzzzxjylrx"
        mode="light"
        shape="roundedSquare"
        title="#b2dfdb"
      />
      <div
        class="sc-bcXHqe junUgb simpleui-colorpicker-colorselector"
        color="#ffcdd2"
        data-testid="test-colorselector"
        id="4fzzzxjylrx"
        mode="light"
        shape="roundedSquare"
        title="#ffcdd2"
      />
      <div
        class="sc-bcXHqe inHKGm simpleui-colorpicker-colorselector"
        color="#b3e5fc"
        data-testid="test-colorselector"
        id="4fzzzxjylrx"
        mode="light"
        shape="roundedSquare"
        title="#b3e5fc"
      />
      <div
        class="sc-bcXHqe bRjGSW simpleui-colorpicker-colorselector"
        color="#d7ccc8"
        data-testid="test-colorselector"
        id="4fzzzxjylrx"
        mode="light"
        shape="roundedSquare"
        title="#d7ccc8"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ColorList should properly render ColorList component - shape square 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-dkrFOg bgZgCB simpleui-colorpicker-list"
        id="4fzzzxjylrx"
        mode="light"
      >
        <div
          class="sc-bcXHqe kiaURt simpleui-colorpicker-colorselector"
          color="#b2dfdb"
          data-testid="test-colorselector"
          id="4fzzzxjylrx"
          mode="light"
          shape="square"
          title="#b2dfdb"
        />
        <div
          class="sc-bcXHqe cmauoY simpleui-colorpicker-colorselector"
          color="#ffcdd2"
          data-testid="test-colorselector"
          id="4fzzzxjylrx"
          mode="light"
          shape="square"
          title="#ffcdd2"
        />
        <div
          class="sc-bcXHqe fOHpyp simpleui-colorpicker-colorselector"
          color="#b3e5fc"
          data-testid="test-colorselector"
          id="4fzzzxjylrx"
          mode="light"
          shape="square"
          title="#b3e5fc"
        />
        <div
          class="sc-bcXHqe etZazF simpleui-colorpicker-colorselector"
          color="#d7ccc8"
          data-testid="test-colorselector"
          id="4fzzzxjylrx"
          mode="light"
          shape="square"
          title="#d7ccc8"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-dkrFOg bgZgCB simpleui-colorpicker-list"
      id="4fzzzxjylrx"
      mode="light"
    >
      <div
        class="sc-bcXHqe kiaURt simpleui-colorpicker-colorselector"
        color="#b2dfdb"
        data-testid="test-colorselector"
        id="4fzzzxjylrx"
        mode="light"
        shape="square"
        title="#b2dfdb"
      />
      <div
        class="sc-bcXHqe cmauoY simpleui-colorpicker-colorselector"
        color="#ffcdd2"
        data-testid="test-colorselector"
        id="4fzzzxjylrx"
        mode="light"
        shape="square"
        title="#ffcdd2"
      />
      <div
        class="sc-bcXHqe fOHpyp simpleui-colorpicker-colorselector"
        color="#b3e5fc"
        data-testid="test-colorselector"
        id="4fzzzxjylrx"
        mode="light"
        shape="square"
        title="#b3e5fc"
      />
      <div
        class="sc-bcXHqe etZazF simpleui-colorpicker-colorselector"
        color="#d7ccc8"
        data-testid="test-colorselector"
        id="4fzzzxjylrx"
        mode="light"
        shape="square"
        title="#d7ccc8"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`ColorList should properly render ColorList component with EyeDropper 1`] = `
<div>
  <div
    class="sc-dkrFOg bgZgCB simpleui-colorpicker-list"
    id="4fzzzxjylrx"
    mode="light"
  >
    <div
      class="sc-bcXHqe evoygh simpleui-colorpicker-colorselector"
      color="#b2dfdb"
      data-testid="test-colorselector"
      id="4fzzzxjylrx"
      mode="light"
      shape="circle"
      title="#b2dfdb"
    />
    <div
      class="sc-bcXHqe jxHNVE simpleui-colorpicker-colorselector"
      color="#ffcdd2"
      data-testid="test-colorselector"
      id="4fzzzxjylrx"
      mode="light"
      shape="circle"
      title="#ffcdd2"
    />
    <div
      class="sc-bcXHqe ldIywx simpleui-colorpicker-colorselector"
      color="#b3e5fc"
      data-testid="test-colorselector"
      id="4fzzzxjylrx"
      mode="light"
      shape="circle"
      title="#b3e5fc"
    />
    <div
      class="sc-bcXHqe jRrVxh simpleui-colorpicker-colorselector"
      color="#d7ccc8"
      data-testid="test-colorselector"
      id="4fzzzxjylrx"
      mode="light"
      shape="circle"
      title="#d7ccc8"
    />
    <div
      class="sc-gswNZR gQYeJH simpleui-colorpicker-eyedropper"
      data-testid="test-eyedropper"
      id="4fzzzxjylrx"
      mode="light"
      shape="circle"
      title="EyeDropper"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M50.75 333.25c-12 12-18.75 28.28-18.75 45.26V424L0 480l32 32 56-32h45.49c16.97 0 33.25-6.74 45.25-18.74l126.64-126.62-128-128L50.75 333.25zM483.88 28.12c-37.47-37.5-98.28-37.5-135.75 0l-77.09 77.09-13.1-13.1c-9.44-9.44-24.65-9.31-33.94 0l-40.97 40.97c-9.37 9.37-9.37 24.57 0 33.94l161.94 161.94c9.44 9.44 24.65 9.31 33.94 0L419.88 288c9.37-9.37 9.37-24.57 0-33.94l-13.1-13.1 77.09-77.09c37.51-37.48 37.51-98.26.01-135.75z"
        />
      </svg>
    </div>
  </div>
</div>
`;
